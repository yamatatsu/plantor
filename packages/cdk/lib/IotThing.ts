import * as cdk from "aws-cdk-lib";
import * as iot from "aws-cdk-lib/aws-iot";

type Props = cdk.StackProps & {
  clientCertArn: string;
};

export class IotThing extends cdk.Stack {
  constructor(parent: cdk.App, id: string, props: Props) {
    super(parent, id, props);

    const { clientCertArn } = props;

    const thing = new iot.CfnThing(this, "Thing", {
      thingName: "wio_terminal_01",
      attributePayload: {},
    });

    const policy = new iot.CfnPolicy(this, "Policy", {
      policyName: "wio_terminal_policy",
      policyDocument: {
        Version: "2012-10-17",
        Statement: [
          {
            Effect: "Allow",
            Action: [
              "iot:Connect",
              "iot:Receive",
              "iot:Publish",
              "iot:Subscribe",
            ],
            /**
             * You can get higher security if you set Client ARNs instead of `*`.
             * @see https://docs.aws.amazon.com/iot/latest/developerguide/create-iot-resources.html#create-iot-policy
             */
            Resource: "*",
          },
        ],
      },
    });

    /**
     * これにより、証明書が権限を持つ。（認可）
     * この設定がないと、証明書があってもmqttとかは認証エラーとなる。
     */
    new iot.CfnPolicyPrincipalAttachment(this, "PPA", {
      policyName: policy.policyName!,
      principal: clientCertArn,
    });

    /**
     * これにより、証明書が thing とマッピングする。
     * この設定がないと AWS IoT は mqtt を受け取っても各機能に紐付けられない。
     */
    new iot.CfnThingPrincipalAttachment(this, "TPA", {
      thingName: thing.thingName!,
      principal: clientCertArn,
    });
  }
}
