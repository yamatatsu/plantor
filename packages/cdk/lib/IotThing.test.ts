import { SynthUtils } from "@aws-cdk/assert";
import { App } from "aws-cdk-lib";
import { IotThing } from "./IotThing";

test("snapshot test", () => {
  const app = new App();

  const target = new IotThing(app, "Target", {
    clientCertArn: "test-clientCertArn",
  });

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot();
});
