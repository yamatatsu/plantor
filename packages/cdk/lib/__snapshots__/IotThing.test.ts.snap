// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Resources": Object {
    "PPA": Object {
      "DependsOn": Array [
        "Policy",
      ],
      "Properties": Object {
        "PolicyName": "wio_terminal_policy",
        "Principal": "test-clientCertArn",
      },
      "Type": "AWS::IoT::PolicyPrincipalAttachment",
    },
    "Policy": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iot:Connect",
                "iot:Receive",
                "iot:Publish",
                "iot:Subscribe",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "wio_terminal_policy",
      },
      "Type": "AWS::IoT::Policy",
    },
    "TPA": Object {
      "DependsOn": Array [
        "Thing",
      ],
      "Properties": Object {
        "Principal": "test-clientCertArn",
        "ThingName": "wio_terminal_01",
      },
      "Type": "AWS::IoT::ThingPrincipalAttachment",
    },
    "Thing": Object {
      "Properties": Object {
        "AttributePayload": Object {},
        "ThingName": "wio_terminal_01",
      },
      "Type": "AWS::IoT::Thing",
    },
  },
}
`;
